### PR TITLE
Remove tactic `hresolve_core`

### LINKED ISSUE #16846: Are `hresolve_core` and `hget_evar` still useful?
`resolve_core` and `hget_evar` were added in 028cbb32785b559c637f77864ce5172e0255d0d0.

This webpage by @gilhur explains their purpose: https://sf.snu.ac.kr/gil.hur/Hpattern/readme.txt

Is this library still working / still used today and still relying on these internal tactics? If yes, we could document these tactics as internal and reference the use case. But we should also consider whether `hpattern` should be distributed more widely by being put into an opam package or even in the standard library.

### PR DESCRIPTION
**Pull request by SkySkimmer**
The implementation is fragile due to being based on detyping and a loc hack, and it seems unused.

It was also not in the doc until
https://github.com/coq/coq/pull/16498 (8.17) and then only got a doc stub.

Overlays:
- https://github.com/snu-sf/paco/pull/47 (note this is just a removal of unused tactic definitions)

### CODE REVIEW

**Comment by SkySkimmer** [comment]
@coqbot run full ci

**Comment by JasonGross** [comment]
According to https://github.com/coq/coq/issues/16846 it is used to implement `hpattern` which looks like a useful tactic that I'd be interested in seeing upstreamed into Coq rather than broken

**Comment by SkySkimmer** [comment]
@coqbot run full ci

**Comment by Zimmi48** [comment]
Given why `hresolve_core` was introduced, I don't think it should be removed without `hget_evar` being removed at the same time.

File: doc/tools/docgram/fullGrammar [hunk 1]
```diff
@@ -1665,8 +1665,6 @@ simple_tactic: [
 | "generalize_eqs_vars" hyp
 | "dependent" "generalize_eqs_vars" hyp
 | "specialize_eqs" hyp
-| "hresolve_core" "(" ident ":=" constr ")" "at" nat_or_var "in" constr
-| "hresolve_core" "(" ident ":=" constr ")" "in" constr
 | "hget_evar" nat_or_var
```
Comment (ppedrot) [inline doc/tools/docgram/fullGrammar:1668]: This should be removed, and similarly in orderedGrammar.

Comment (jfehrle) [reply doc/tools/docgram/fullGrammar:1668]: Run `make doc_gram_rst` to update this.  (You failed lint for the lack of this step).

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
@coqbot merge now

**Comment by herbelin** [comment]
> According to https://github.com/coq/coq/issues/16846 it is used to implement hpattern which looks like a useful tactic that I'd be interested in seeing upstreamed into Coq rather than broken

Ideas from `hresolve_core` were used in `destruct` (via `Evarconv.second_order_matching`). I don't remember the exact differences between `Evarconv.second_order_matching` and `hresolve_core`, but for sure, these are ideas to valorize and to develop further. Iirc, Matthieu was working on generalizing the use of  `Evarconv.second_order_matching` as part of Unifall, but, again, I don't know the exact status of it.

**Pull request merged**